Add dynamic table log writer

achulkov2 opened the pull request:
Adds a new `dynamic_table` log writer, which writes log events to the configured sorted or ordered dynamic table.
Log events are serialized into a YSON representation, which is interpreted similarly to the `insert-rows` cli command.
It is your responsibility to create the table with a compatible schema. 
See integration test for example configuration.

The writer is available for use in all YT server components.

gritukan commented on yt/yt/client/logging/dynamic_table_log_writer.cpp:
```diff
@@ -0,0 +1,372 @@
+
+        // This is necessary to avoid hitting the limit of allowed row modifications within a single transaction.
+        // Also, there is no upside from huge batches.
+        if (CurrentRowCount_ > Config_->MaxBatchRowCount || std::ssize(CurrentBuffer_) > Config_->MaxBatchWeight) {
```
Technically >=?

gritukan commented on yt/yt/client/logging/dynamic_table_log_writer.cpp:
```diff
@@ -0,0 +1,372 @@
+        , Formatter_(std::move(formatter))
+        , Config_(config)
+        , Invoker_(std::move(invoker))
+    , FlushExecutor_(New<TPeriodicExecutor>(
```
indentation

gritukan commented on yt/yt/client/logging/dynamic_table_log_writer.cpp:
```diff
@@ -0,0 +1,372 @@
+        }
+
+        auto tableMountCache = client->GetTableMountCache();
+        auto tableInfo = WaitFor(tableMountCache->GetTableInfo(Config_->TablePath))
```
WaitForFast can be done here to avoid a context switch in a typical case. It does not make much sense in this case, but is a good pattern anyway

gritukan commented on yt/yt/client/logging/dynamic_table_log_writer.cpp:
```diff
@@ -0,0 +1,372 @@
+            RotateBuffer();
+        }
+
+        TStringOutput outputStream(CurrentBuffer_);
```
Why do we convert batch of events to a single string first and then convert them into a list of rows? Is it possible to create rows here?

achulkov2 replied:
The main problem is that we need to know the table schema in order to convert the yson strings into unversioned rows, and getting the schema is a YT client request.

The tricky part is that the schema might change, for example, we forgot to add a column, saw the errors in logs and then altered the table to contain the forgotten column.

So either we need some periodic that will fetch and store the schema, or we fetch it upon write.

For now I decided to go with the second approach. Later on, if we switch to some writer that needs to know the schema upon creation, we can switch to the first approach and move parsing into `Write`. However, the writer will have to support schema-alerting :)

achulkov2 replied:
This is considered resolved (without any changes), but I will keep this thread open for now, so that future reviewers see it.

gritukan commented on yt/yt/core/logging/log_writer_detail.cpp:
```diff
@@ -73,50 +73,131 @@ i64 TRateLimitCounter::GetAndResetLastSkippedEventsCount()
+
+void TRateLimitingLogWriterBase::WriteLogSkippedEvent(i64 eventsSkipped, TStringBuf skippedBy)
+{
+    if (eventsSkipped > 0) {
```
These ifs can be joined with &&

achulkov2 commented on yt/yt/tests/integration/misc/test_logging.py:
```diff
@@ -0,0 +1,166 @@
+            # This is specified to test rate limits.
+            "category_rate_limits": {
+                # No sequoia logs for you!
+                "SequoiaServer": 0,
```
It may be wise to use another category, since `Sequoia` logs are relatively new and are not present in 23.2, for example.

achulkov2 replied:
Changed to Concurrency.

achulkov2 commented on yt/yt/core/logging/formatter.cpp:
```diff
@@ -86,10 +90,35 @@ TLogEvent GetSkippedLogStructuredEvent(i64 count, TStringBuf skippedBy)
+////////////////////////////////////////////////////////////////////////////////
+
+TPlainTextLogFormatter::TPlainTextLogFormatter(bool enableSourceLocation)
+    : TLogFormatterBase( enableSourceLocation)
```
Extra space.

maxim-babenko commented on yt/yt/client/logging/dynamic_table_log_writer.cpp:
```diff
@@ -0,0 +1,373 @@
+// This log category is special in the default log manager: events with this category are logged solely to stderr.
+// Note that this behaviour does not depend on the log level of the message.
+// We use this logger to (try our best) to avoid perpetual log message production.
+// Currently it is impossible to disable logging along a chain of invocations completely, so some induced log messages will loop back into our writer.
```
Some lines here are way too long, let's split them

maxim-babenko commented on yt/yt/client/logging/dynamic_table_log_writer.cpp:
```diff
@@ -0,0 +1,373 @@
+    {
+        VERIFY_INVOKER_AFFINITY(Invoker_);
+
+        TString ysonRows;
```
Why TString, not TYsonString?
Or maybe even shared refs to ease memory tracking?

achulkov2 replied:
Changed to `TSharedRef` for better memory tracking. But more generally speaking I would prefer to move towards the ability to pass yson input to C++ client write methods.

achulkov2 replied:
Switched to shared refs.

maxim-babenko commented on yt/yt/client/logging/dynamic_table_log_writer.cpp:
```diff
@@ -0,0 +1,373 @@
+        auto config = ParseConfig(configNode);
+        if (config->Format != ELogFormat::Yson) {
+            THROW_ERROR_EXCEPTION(
+                "Writers of type %Qv must be configured to use %Qv format",
```
Qlv for format

maxim-babenko commented on yt/yt/client/logging/dynamic_table_log_writer.cpp:
```diff
@@ -0,0 +1,373 @@
+public:
+    TDynamicTableLogWriterFactory()
+        : ClientHolder_(New<TClientHolder>())
+        , WriteThreadPool_(CreateThreadPool(1, "LogYTWrite"))
```
Do we need a thread pool here?

achulkov2 replied:
There is some serialization/deserialization load, but I guess we can live with a single action queue thread for now. If necessary, we can change it to a thread pool with a configurable number of threads.

achulkov2 commented on yt/yt/client/logging/dynamic_table_log_writer.cpp:
```diff
@@ -0,0 +1,373 @@
+    // Moreover, the current log manager is not designed for handling asynchronous event
+    // writers. All write/flush calls are made consecutively and are considered successful
+    // at once. Changing this would require modifying the logging event pipeline significantly.
+    i64 WriteImpl(const TLogEvent& event)
```
Moved to private here and in log_writer_detail.h.

achulkov2 commented on yt/yt/client/ya.make:
```diff
@@ -180,6 +180,9 @@ SRCS(
     complex_types/uuid_text.cpp
     complex_types/time_text.cpp
 
+    logging/config.cpp
```
Sure, why not. It still depends on `yt/yt/client` though.

achulkov2 commented:
Addressed all comments by @maxim-babenko and added some profiling which we found necessary after deploying this feature locally. Placed these changes in a separate commit for easier reading.

eshcherbin commented on yt/yt/core/logging/formatter.cpp:
```diff
@@ -205,6 +213,9 @@ i64 TStructuredLogFormatter::WriteFormatted(IOutputStream* stream, const TLogEve
             })
+            .DoIf(EnableHostField_, [&] (auto fluent) {
+                fluent.Item("host").Value(NNet::GetLocalHostName());
+            })
```
Could you explain the need for this "host" field?
I don't like having it separated from the other system fields. If we want to be able to enable only a subset of system fields, I'd suggest changing the bool option with a flag enum. However, it should be done in a separate PR, this one's already complex enough.

achulkov2 replied:
It is useful when you have multiple hosts producing logs to the same dynamic table. While this is something that we could append in the specific implementation for dynamic tables, I think it is closer to the other system fields and might be useful in some other log writer later on. 

I don't like it separated either, but I don't want to break existing structured logs. Changing to a flag enum is a good idea, I'll write a TODO.

eshcherbin commented on yt/yt/core/logging/log_manager.cpp:
```diff
@@ -503,6 +504,11 @@ class TLogManager::TImpl
+        // For now this is the only way to wait for log writers that perform asynchronous flushes.
+        if (Config_->ShutdownBusyTimeout) {
+            Sleep(Config_->ShutdownBusyTimeout);
+        }
```
I don't really like this sleep here. However, I understand that it would require a considerable refactoring to be able to wait for the dynamic table log writer closing. For now, let's at least add a TODO for later.

achulkov2 commented on yt/yt/core/logging/log_writer_detail.cpp:
```diff
@@ -73,50 +73,128 @@ i64 TRateLimitCounter::GetAndResetLastSkippedEventsCount()
-    : Formatter_(std::move(formatter))
+TSegmentSizeReporter::TSegmentSizeReporter(TString logWriterName)
+    : CurrentSegmentSizeGauge_(
+        TProfiler{"/logging"}.WithSparse().WithTag("writer", std::move(logWriterName)).Gauge("/current_segment_size"))
```
I am dangerously scared of using global variables in loggers, because a bunch of the logic happens during static initialization. I had my fun with siof while writing the code. Plus this is code by prime@, whom I tend to trust in these matters :) 

But I will give it some thought in hopes of convincing myself that using a global profiler would be fine.

eshcherbin replied:
OK, no strong opinion here.

achulkov2 replied:
I gave it some thought and decided that I do not want to interfere with the ability to configure `TLogManager` before main. Whether this is actually allowed now is debatable and depends on the current release (for example, loggers themselves became siof-safe in trunk). To be on the safer side while still introducing some cleanliness, I moved the profiler into a class field.

achulkov2 commented on yt/yt/core/logging/log_writer_detail.cpp:
```diff
@@ -73,50 +73,128 @@ i64 TRateLimitCounter::GetAndResetLastSkippedEventsCount()
+    if (!SystemCategory_) {
+        SystemCategory_ = GetDefaultLogManager()->GetCategory(SystemLoggingCategoryName);
+    }
+    bool isSystemEvent = event.Category == SystemCategory_;
```
It is new, it seemed reasonable from a general perspective when I was writing the code.

Now that I think about it, I can't say it makes too much of a difference: log writers that would receive system category events won't actually have any limits configured, since they are handled by a designated error log writer anyway. Meanwhile, log skipped system events will get written in any case, since they are not looped back into the rate limiting logic. 

I am fine with reverting this part, if you feel like it would improve readability.

eshcherbin replied:
I'm not really worried about readability here, I just couldn't verify that this is 100% safe with a quick glance (and I couldn't allocate more time to think about it, given that it's not essential for the PR's primary goal).

Not really a strong opinion, but reverting it would make the code a bit tidier, yes.

achulkov2 replied:
Reverted this part.

achulkov2 commented on yt/yt/core/logging/log_writer_detail.h:
```diff
@@ -38,37 +40,90 @@ class TRateLimitCounter
+private:
+    std::atomic<i64> CurrentSegmentSize_ = 0;
+    NProfiling::TGauge CurrentSegmentSizeGauge_;
+};
```
This class got separated during the many attempts of making this whole construction work, it is entirely possible that it doesn't need to be a separate class anymore (as it was originally). I will give it another look.

achulkov2 replied:
Moved this logic back into `TRateLimitingLogWriterBase` as protected methods.

maxim-babenko commented:
Shipped!

achulkov2 commented on yt/yt/client/logging/dynamic_table_log_writer.cpp:
```diff
@@ -0,0 +1,422 @@
+
+        ++CurrentRowCount_;
+        BacklogEventCount_.fetch_add(1, std::memory_order::relaxed);
+        BacklogEventsGauge_.Update(BacklogEventCount_.load(std::memory_order::relaxed));
```
Cool, didn't know about this functionality before!

maxim-babenko commented:
Shipped!
